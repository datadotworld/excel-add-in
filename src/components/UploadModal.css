/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
 .upload-modal {
    top: 50px;
  }
  
  .selection {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
  
  .selection span {
    margin: 0;
    padding-left: 0.125rem;
    height: 30px;
    display: flex;
    align-items: center;
    font-weight: normal;
    font-size: 0.875rem;
    font-weight: 700;
  }
  
  .selection span::before {
    content: '';
    display: inline-block;
    width: 1em;
    height: 1em;
    border-radius: 1em;
    margin-right: 0.75em;
    background: white;
    border: 1px solid lightgray
  }
  
  .selection input[type=radio] {
    display: none;
  }
  
  .selection input[type=radio]:checked + span::before {
    border: 4.5px solid #328df3;
    background: white;
  }
  
  .selection .radio-button {
    display: flex;
    align-items: center;
    cursor: pointer;
  }
  
  .selection .selection-info {
    display: flex;
    padding: 12px;
    color: #969696;
    font-size: 11px;
  }
  
  .select-sheet {
    background-color: aliceblue;
  }
  
  .full-screen-modal .center-block .selection-form {
    padding: 0;
  }

  .url {
      display: flex;
  }

  .url-group {
    border-top: 1px solid lightgray;
    border-bottom: 1px solid lightgray
  }

  .url-group .body {
    margin-top: 1rem;
  }

  .category-title {
    background-color: #fff;
    height: 3rem;
    display: flex;
    align-items: center;
    padding-left: 0.9375rem;
    font-size: 18px;
    border-bottom: 1px solid lightgray;
  }
  .category-reminder {
    font-size: 0.9375rem;
    font-weight: 700;
    padding: 1rem;
    text-align: center;
    color: lightgray;
  }
  
  .category-title .control-label {
    margin-bottom: 0;
  }

  .browse-button {
    width: 25% !important;
    margin-left: 0.625rem;
    padding: 0 0.625rem;
    height: 2rem;
    font-size: 0.8125rem;
    background: white;
  }

  .help-block {
    font-size: 0.6875rem;
    margin-top: 0.3125rem;
  }

  .upload-modal .control-label {
    font-size: 0.9375rem;
  }

  .upload-modal .form-control {
    height: 2rem;
    border-radius: 0.3125rem !important;
  }

  .textField {
    font-size: 0.875rem;
  }

  .cancel-button {
    margin-left: 0 !important;
  }