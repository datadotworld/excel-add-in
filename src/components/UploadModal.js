import React, { Component } from 'react';
import PropTypes from 'prop-types';

import {
  Button,
  Grid,
  Row,
  HelpBlock,
  FormControl,
  InputGroup,
  ControlLabel,
  FormGroup
} from 'react-bootstrap';
import {
  MAX_FILENAME_LENGTH,
  SHEET_RANGE,
  MAX_COLUMNS,
  MAX_COLUMNS_ERROR
} from '../constants';
import { getDisplayRange } from '../util';
import analytics from '../analytics';

import WarningModal from './WarningModal';
import './UploadModal.css';

const getFilenameWithoutExtension = function(filename) {
  const dotPos = filename.lastIndexOf('.');
  return dotPos > -1 ? filename.slice(0, dotPos) : filename;
};

const filenameRegex = /^[^/]+$/;

export default class UploadModal extends Component {
  static propTypes = {
    excelApiSupported: PropTypes.bool,
    range: PropTypes.object
  };

  state = {
    filename: '',
    currentUrl: this.props.url,
    showWarningModal: false
  };

  componentDidMount() {
    // To show number of rows and columns selected
    this.props.getSelectionRange();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.url !== this.props.url) {
      this.setState({ currentUrl: this.props.url });
    }
  }

  isFormValid = () => {
    const { filename } = this.state;
    const { excelApiSupported, range } = this.props;
    if ((excelApiSupported && !range) || !filename) {
      return false;
    }
    return (
      filename.match(filenameRegex) && filename.length < MAX_FILENAME_LENGTH
    );
  };

  getSelectionText(range) {
    if (range) {
      // Return number of rows and columns in selection
      const rowCount = range.rowCount;
      const columnCount = range.columnCount;
      const rowText = rowCount === 1 ? 'Row' : 'Rows';
      const columnText = columnCount === 1 ? 'Column' : 'Columns';

      return `(${rowCount} ${rowText} x ${columnCount} ${columnText})`;
    }

    // Range not yet loaded, return placeholder text
    return '(Rows x Columns)';
  }

  getSheetName = (range) => {
    if (range) {
      const address = range.address;
      // Extract sheet name and trim quote marks which are added when the name has a space
      const sheet = address
        .substring(0, address.indexOf('!'))
        .replace(/'/g, '');
      return sheet;
    }

    return '';
  };

  submitFile = async () => {
    this.closeModal();
    const {
      close,
      sync,
      setError,
      selectSheet,
      getSelectionRange
    } = this.props;
    const { filename } = this.state;

    try {
      const range = await getSelectionRange();
      const { columnCount } = range;

      if (columnCount >= MAX_COLUMNS) {
        this.props.setErrorMessage(MAX_COLUMNS_ERROR);
      } else {
        let rangeAddress = range.address;
        if (rangeAddress) {
          if (selectSheet) {
            const [sheet] = rangeAddress.split('!');
            rangeAddress = `${sheet}!${SHEET_RANGE}`;
          }

          await sync(
            `${filename}.csv`,
            rangeAddress.replace(/'/g, ''),
            this.state.currentUrl,
            range.worksheet.id
          );
          await close();
        }
      }
    } catch (selectionRangeError) {
      setError(selectionRangeError);
    }
  };

  submit = (event) => {
    const { filename } = this.state;
    analytics.track('exceladdin.add_data.submit.click');
    event.preventDefault();

    if (this.props.doesFileExist(`${filename}.csv`)) {
      // Show warning modal
      this.setState({ showWarningModal: true });
    } else {
      this.submitFile();
    }
  };

  getFilename = () => {
    const index = this.state.name.lastIndexOf('.csv');
    if (index >= 0 && index === this.state.name.length - 4) {
      return getFilenameWithoutExtension(this.state.name);
    }
    return this.state.name;
  };

  closeModal = () => {
    this.setState({ showWarningModal: false });
  };

  cancelClicked = () => {
    analytics.track('exceladdin.add_data.cancel.click');
    this.props.close();
  };

  handleUrlChange(url) {
    this.setState({ currentUrl: url });
  }

  render() {
    const {
      excelApiSupported,
      range,
      numItemsInHistory,
      selectSheet
    } = this.props;
    const { filename } = this.state;
    let validState, selection;
    if (filename) {
      validState = this.isFormValid() ? 'success' : 'warning';
    }

    if (selectSheet) {
      selection = range ? this.getSheetName(range) : '';
    } else {
      selection = range ? range.address : '';
    }

    return (
      <div>
        <div className="full-screen-modal category-title">
          <ControlLabel>Upload to data.world</ControlLabel>
        </div>
        <Grid className="upload-modal full-screen-modal  ">
          <Row className="center-block">
            <form onSubmit={this.submit}>
              {excelApiSupported && (
                <FormGroup>
                  <ControlLabel>Select the area or sheet to save:</ControlLabel>
                  <div
                    className="selection-form"
                    onClick={this.props.changeSelection}
                  >
                    <div className="selection">
                      <label className="radio-button">
                        <input
                          type="radio"
                          name="selection"
                          value="selection"
                          checked={!selectSheet}
                          readOnly
                        />
                        <span>Selection</span>
                      </label>
                      <div className="selection-info">
                        {this.getSelectionText(range)}
                      </div>
                    </div>
                    <div className="selection">
                      <label className="radio-button">
                        <input
                          type="radio"
                          name="selection"
                          value="sheet"
                          checked={selectSheet}
                          readOnly
                        />
                        <span>Sheet</span>
                      </label>
                      <div className="selection-info">
                        {`(${this.getSheetName(range)})`}
                      </div>
                    </div>
                  </div>
                </FormGroup>
              )}
              <FormGroup className="url-group" validationState={validState}>
                <div className="body">
                  <ControlLabel>Dataset or project URL:</ControlLabel>
                  <InputGroup>
                    <div className="url">
                      <FormControl
                        className="textField"
                        placeholder="https://data.world/"
                        value={this.state.currentUrl}
                        type="text"
                        onChange={(event) => {
                          this.handleUrlChange(event.target.value);
                        }}
                      />
                      <Button
                        className="browse-button"
                        onClick={() => this.props.showDatasets()}
                      >
                        Browse
                      </Button>
                    </div>
                  </InputGroup>
                  <HelpBlock className="help-block">
                    Copy/paste the URL of a dataset or project, or click
                    "Browse"
                  </HelpBlock>
                </div>
              </FormGroup>
              <FormGroup validationState={validState}>
                <ControlLabel>File name:</ControlLabel>
                <InputGroup>
                  <FormControl
                    value={filename}
                    className="textField"
                    type="text"
                    onChange={({ target }) => {
                      this.setState({ filename: target.value });
                    }}
                  />
                </InputGroup>
                <HelpBlock>
                  Must not include slashes, your file will be named{' '}
                  <strong>
                    {filename}
                    .csv
                  </strong>
                  <div className="titleLimit">max. 128</div>
                </HelpBlock>
              </FormGroup>
              <div className="button-group">
                {numItemsInHistory !== 0 && (
                  <Button
                    className="cancel-button"
                    onClick={this.cancelClicked}
                  >
                    Cancel
                  </Button>
                )}
                {this.props.loading ? (
                  <Button
                    type="submit"
                    className={
                      numItemsInHistory !== 0
                        ? 'submit-button'
                        : 'full-submit-button'
                    }
                    disabled
                    bsStyle="primary"
                  >
                    Saving...
                  </Button>
                ) : (
                  <Button
                    type="submit"
                    className={
                      numItemsInHistory !== 0
                        ? 'submit-button'
                        : 'full-submit-button'
                    }
                    bsStyle="primary"
                  >
                    Save file
                  </Button>
                )}
              </div>
            </form>
          </Row>
          <WarningModal
            show={this.state.showWarningModal}
            cancelHandler={this.closeModal}
            successHandler={this.submitFile}
            analyticsLocation="exceladdin.add_data"
          >
            <div>
              <strong>
                "{filename}
                .csv" already exists on data.world. Do you want to replace it?
              </strong>
            </div>
            <div>
              Replacing it will overwrite the file on data.world with the
              contents from{' '}
              {excelApiSupported
                ? getDisplayRange(selection)
                : 'the selected cell range'}
            </div>
          </WarningModal>
        </Grid>
      </div>
    );
  }
}
