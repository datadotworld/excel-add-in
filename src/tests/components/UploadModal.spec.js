/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import UploadModal from '../../components/UploadModal';

it('renders modal', () => {
  expect(renderer.create(
    <UploadModal
      getSelectionRange = {
        () => {
          return new Promise((resolve) => {
            resolve(1);
          });
        }
      }
    />
    ).toJSON()).toMatchSnapshot()
});

it('renders modal - with range', () => {
  expect(renderer.create(
    <UploadModal
      range={{
        rowCount: 4,
        columnCount: 2,
        address: 'Sheet1!A2:B5'
      }}
      excelApiSupported
      linkConnector = {() => {}}
      getSelectionRange = {
        () => {
          return new Promise((resolve) => {
            resolve(1);
          });
        }
      }
      
    />
  ).toJSON()).toMatchSnapshot()
});

it('renders modal - with options', () => {
  const options = {
    filename: 'test1.csv'
  };
  expect(renderer.create(
    <UploadModal
      range={{
        rowCount: 4,
        columnCount: 2,
        address: 'Sheet1!A2:B5'
      }}
      options={options}
      excelApiSupported
      linkConnector = {() => {}}
      getSelectionRange = {
        () => {
          return new Promise((resolve) => {
            resolve();
          });
        }
      }
    />
  ).toJSON()).toMatchSnapshot()
});

it ('removes trailing .csv from filenames', () => {
  const modal = new UploadModal({options: {}});
  modal.state.name = 'testing.csv'
  expect(modal.getFilename()).toBe('testing');
});

it ('does not impact filenames without trailing .csv', () => {
  const modal = new UploadModal({options: {}});
  modal.state.name = 'testingwithoutextension'
  expect(modal.getFilename()).toBe('testingwithoutextension');
});