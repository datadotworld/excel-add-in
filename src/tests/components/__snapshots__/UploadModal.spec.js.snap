// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders modal - with options 1`] = `
<div>
  <div
    className="full-screen-modal category-title"
  >
    <label
      className="control-label"
      htmlFor={undefined}
    >
      Upload to data.world
    </label>
  </div>
  <div
    className="upload-modal full-screen-modal   container"
  >
    <div
      className="center-block row"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor={undefined}
          >
            Select the area or sheet to save:
          </label>
          <div
            className="selection-form"
            onClick={undefined}
          >
            <div
              className="selection"
            >
              <label
                className="radio-button"
              >
                <input
                  checked={true}
                  name="selection"
                  readOnly={true}
                  type="radio"
                  value="selection"
                />
                <span>
                  Selection
                </span>
              </label>
              <div
                className="selection-info"
              >
                (4 Rows x 2 Columns)
              </div>
            </div>
            <div
              className="selection"
            >
              <label
                className="radio-button"
              >
                <input
                  checked={undefined}
                  name="selection"
                  readOnly={true}
                  type="radio"
                  value="sheet"
                />
                <span>
                  Sheet
                </span>
              </label>
              <div
                className="selection-info"
              >
                (Sheet1)
              </div>
            </div>
          </div>
        </div>
        <div
          className="url-group form-group"
        >
          <div
            className="body"
          >
            <label
              className="control-label"
              htmlFor={undefined}
            >
              Dataset or project URL:
            </label>
            <span
              className="input-group"
            >
              <div
                className="url"
              >
                <input
                  className="textField form-control"
                  id={undefined}
                  onChange={[Function]}
                  placeholder="https://data.world/"
                  type="text"
                  value={undefined}
                />
                <button
                  className="browse-button btn btn-default"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Browse
                </button>
              </div>
            </span>
            <span
              className="help-block help-block"
            >
              Copy/paste the URL of a dataset or project, or click "Browse"
            </span>
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor={undefined}
          >
            File name:
          </label>
          <span
            className="input-group"
          >
            <input
              className="textField form-control"
              id={undefined}
              onChange={[Function]}
              type="text"
              value=""
            />
          </span>
          <span
            className="help-block"
          >
            Must not include slashes, your file will be named 
            <strong>
              
              .csv
            </strong>
            <div
              className="titleLimit"
            >
              max. 128
            </div>
          </span>
        </div>
        <div
          className="button-group"
        >
          <button
            className="cancel-button btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
          <button
            className="submit-button btn btn-primary"
            disabled={true}
            type="submit"
          >
            Save file
          </button>
        </div>
      </form>
    </div>
    <div
      className="static-modal"
    />
  </div>
</div>
`;

exports[`renders modal - with range 1`] = `
<div>
  <div
    className="full-screen-modal category-title"
  >
    <label
      className="control-label"
      htmlFor={undefined}
    >
      Upload to data.world
    </label>
  </div>
  <div
    className="upload-modal full-screen-modal   container"
  >
    <div
      className="center-block row"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor={undefined}
          >
            Select the area or sheet to save:
          </label>
          <div
            className="selection-form"
            onClick={undefined}
          >
            <div
              className="selection"
            >
              <label
                className="radio-button"
              >
                <input
                  checked={true}
                  name="selection"
                  readOnly={true}
                  type="radio"
                  value="selection"
                />
                <span>
                  Selection
                </span>
              </label>
              <div
                className="selection-info"
              >
                (4 Rows x 2 Columns)
              </div>
            </div>
            <div
              className="selection"
            >
              <label
                className="radio-button"
              >
                <input
                  checked={undefined}
                  name="selection"
                  readOnly={true}
                  type="radio"
                  value="sheet"
                />
                <span>
                  Sheet
                </span>
              </label>
              <div
                className="selection-info"
              >
                (Sheet1)
              </div>
            </div>
          </div>
        </div>
        <div
          className="url-group form-group"
        >
          <div
            className="body"
          >
            <label
              className="control-label"
              htmlFor={undefined}
            >
              Dataset or project URL:
            </label>
            <span
              className="input-group"
            >
              <div
                className="url"
              >
                <input
                  className="textField form-control"
                  id={undefined}
                  onChange={[Function]}
                  placeholder="https://data.world/"
                  type="text"
                  value={undefined}
                />
                <button
                  className="browse-button btn btn-default"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Browse
                </button>
              </div>
            </span>
            <span
              className="help-block help-block"
            >
              Copy/paste the URL of a dataset or project, or click "Browse"
            </span>
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor={undefined}
          >
            File name:
          </label>
          <span
            className="input-group"
          >
            <input
              className="textField form-control"
              id={undefined}
              onChange={[Function]}
              type="text"
              value=""
            />
          </span>
          <span
            className="help-block"
          >
            Must not include slashes, your file will be named 
            <strong>
              
              .csv
            </strong>
            <div
              className="titleLimit"
            >
              max. 128
            </div>
          </span>
        </div>
        <div
          className="button-group"
        >
          <button
            className="cancel-button btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
          <button
            className="submit-button btn btn-primary"
            disabled={true}
            type="submit"
          >
            Save file
          </button>
        </div>
      </form>
    </div>
    <div
      className="static-modal"
    />
  </div>
</div>
`;

exports[`renders modal 1`] = `
<div>
  <div
    className="full-screen-modal category-title"
  >
    <label
      className="control-label"
      htmlFor={undefined}
    >
      Upload to data.world
    </label>
  </div>
  <div
    className="upload-modal full-screen-modal   container"
  >
    <div
      className="center-block row"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="url-group form-group"
        >
          <div
            className="body"
          >
            <label
              className="control-label"
              htmlFor={undefined}
            >
              Dataset or project URL:
            </label>
            <span
              className="input-group"
            >
              <div
                className="url"
              >
                <input
                  className="textField form-control"
                  id={undefined}
                  onChange={[Function]}
                  placeholder="https://data.world/"
                  type="text"
                  value={undefined}
                />
                <button
                  className="browse-button btn btn-default"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Browse
                </button>
              </div>
            </span>
            <span
              className="help-block help-block"
            >
              Copy/paste the URL of a dataset or project, or click "Browse"
            </span>
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor={undefined}
          >
            File name:
          </label>
          <span
            className="input-group"
          >
            <input
              className="textField form-control"
              id={undefined}
              onChange={[Function]}
              type="text"
              value=""
            />
          </span>
          <span
            className="help-block"
          >
            Must not include slashes, your file will be named 
            <strong>
              
              .csv
            </strong>
            <div
              className="titleLimit"
            >
              max. 128
            </div>
          </span>
        </div>
        <div
          className="button-group"
        >
          <button
            className="cancel-button btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
          <button
            className="submit-button btn btn-primary"
            disabled={true}
            type="submit"
          >
            Save file
          </button>
        </div>
      </form>
    </div>
    <div
      className="static-modal"
    />
  </div>
</div>
`;
